payload/Library/nudge/Resources/nudgelib/ux.py

```diff
@@ -0,0 +1,209 @@
+                      get_serial,
+                      pending_apple_updates,
+                      nudgelog)
+from .prefs import set_app_pref
+
+
+class UX():
+    '''Class that define the UX a user will have'''
+    def __init__(self, builder_obj, nudge_prefs):
+        self.nibbler = builder_obj
+        self.nudge = self.nibbler.nudge
+        self.nudge_path = self.nibbler.nudge_path
+        self.nudge_prefs = nudge_prefs
+        self.date_diff_seconds = 0
+        self.date_diff_days = 0
+        self.cut_off_warn = False
+        self.timer = 0
+
+    def build(self):
+        '''Function that will build the UX for the user'''
+        self._set_images()
+        self._set_buttons()
+        self._set_static_fields()
+        self._set_dynamic_fields()
+        self._hide_more_info()
+        self._set_timer_logic()
+
+    def run(self):
+        '''Function that will run nudge experience'''
+        set_app_pref('last_seen', NSDate.new())
+        self.nudge.hidden = True
+        self.nudge.run()
+
+    def _set_images(self):
+        for index, path in enumerate([self.nudge_prefs['logo_path'],
+                                      self.nudge_prefs['screenshot_path']]):
+            if path in DEFAULT_IMAGES:
+                local_png_path = join(
+                    self.nudge_path, path).replace(' ', '%20')
+            else:
+                local_png_path = join(path).replace(' ', '%20')
+            foundation_nsurl_path = NSURL.URLWithString_(
+                f'file:{local_png_path}')
+            foundation_nsdata = NSData.dataWithContentsOfURL_(
+                foundation_nsurl_path)
+            foundation_nsimage = NSImage.alloc().initWithData_(
+                foundation_nsdata)
+            if index == 0:
+                self.nudge.views[COMPANY_LOGO].setImage_(foundation_nsimage)
+            else:
+                self.nudge.views[UPDATESS].setImage_(foundation_nsimage)
+
+    def _set_buttons(self):
+        self.nudge.attach(self.nibbler.button_update, 'button.update')
+        self.nudge.attach(self.nibbler.button_moreinfo, 'button.moreinfo')
+        self.nudge.attach(self.nibbler.button_ok, 'button.ok')
+        self.nudge.attach(self.nibbler.button_understand, 'button.understand')
+
+    def _set_static_fields(self):
+        ui_nudge_prefs = iter(UI_NUDGE_PREFS)
+        [self.nudge.views[field].setStringValue_(
+            self.nudge_prefs[next(ui_nudge_prefs)]) for field in UI_FIELDS]
+
+    def _set_dynamic_fields(self):
+        self.nudge.views[USERNAME].setStringValue_(str(get_console_username()))
+        self.nudge.views[SERIAL_NUMBER].setStringValue_(str(get_serial()))
+        self.nudge.views[UPDATED].setStringValue_('No')
+        self.nudge.views[DEFERRAL_COUNT].setStringValue_(str(NUDGE_DISMISSED_COUNT))
+
+    def _hide_more_info(self):
+        if not self.nudge_prefs['more_info_url']:
+            self.nudge.views[MORE_INFO].setHidden_(True)
+
+    def _set_timer_logic(self):
```

Comment:
As we discussed over Slack, @LcTrKiD, if we could catch errors generated from an ill-configured mobileconfig/JSON that would be great. Specifically to this function, this will throw the following error if someone tries to configure a `cut_off_date` that is not a real day (like I was trying to do when setting it to `2020-09-31`):
```
2020-09-24 14:35:21.672 Python[13805:3351484] [Nudge] OS version is below the minimum threshold subversion: 19G2021
Traceback (most recent call last):
  File "/Library/nudge/Resources/nudge", line 55, in <module>
    main()
  File "/Library/nudge/Resources/nudge", line 50, in main
    ux.build()
  File "/Library/nudge/Resources/nudgelib/ux.py", line 46, in build
    self._set_timer_logic()
  File "/Library/nudge/Resources/nudgelib/ux.py", line 97, in _set_timer_logic
    self._get_date_diff_days()
  File "/Library/nudge/Resources/nudgelib/ux.py", line 115, in _get_date_diff_days
    cut_off_date_strp = datetime.strptime(cut_off_date,
  File "/Library/nudge/Python.framework/Versions/3.8/lib/python3.8/_strptime.py", line 568, in _strptime_datetime
    tt, fraction, gmtoff_fraction = _strptime(data_string, format)
  File "/Library/nudge/Python.framework/Versions/3.8/lib/python3.8/_strptime.py", line 534, in _strptime
    julian = datetime_date(year, month, day).toordinal() - \
ValueError: day is out of range for month
```

Reply:
fixed in https://github.com/macadmins/nudge/pull/49/commits/ab7b5a8062b358c4bf8fa3d5d1d80d360e25e265